### PR TITLE
[image][ios] Fix freeze/delay in rendering of WEBP images on iOS, add controls for stoping/starting animated images

### PR DESCRIPTION
**Pull request by gkasdorf**
# Why

Currently, there is no way to control whether an image such as a GIF or WEBP is animated or not. 

In addition to this, the issue seen in [this issue](https://github.com/expo/expo/issues/24557) regarding freezing/high memory usage during the rendering of WEBP images is fixed.

# How

This PR adds the ability to set an image to either automatically animate animated images (default behavior remains the same as true) and adds two functions to the component, `startAnimating()` and `startAnimating()`.

The issue with freezing is resolved by using `SDAnimatedImage` over `UIImage` for images that are animated on iOS. There is a referenced WEBP file in the issue that I used to test the current behavior vs the behavior with SDAnimatedImage.

Currently, when attempting to load that image we end up with a nearly one minute hang. The app is unresponsive during that time and the FPS drops to zero. Using SDAnimatedImage, we instead get a normal load time as we expect.

We were also unable currently to control the animation behavior of the image, so using SDAnimatedImage allows us to use those controls. There is more functionality that could be added in the future if anyone desired (such as playback speed, direction, as well as an advanced player it appears).

From what I understand SDAnimatedImage extends UIImage, so all of the current behavior should remain the same in terms of resizing, scaling, etc. Non-animated images still use UIImage since some non-animated images will not load otherwise.

I am not overly familiar with Glide but am happy to implement the functionality on Android as well if anyone points me in the correct direction on that. If not, I will look more into it whenever I have some time. I'm also happy to create another PR with only the SDAnimatedImage fix if that is preferred to adding the functionality here.

# Test Plan

Originally, we would encounter situations like this video when attempting to load some WEBP files.

https://github.com/expo/expo/assets/127079571/1e88b6ce-82c3-430b-91bb-339943e1cc87

Now, we see that the images are loaded as soon as the request is complete. We also see that the images can still be re-scaled, non-animated images still render, and multiple image formats are supported.

https://github.com/expo/expo/assets/127079571/d621ceef-8b50-428c-9136-de02d39c943f

I've also added one test that checks the `onLoad` event returns `animated` as `true`. Other tests here still fail, but those were fixed [here](https://github.com/expo/expo/pull/24980) and I don't want for there to be au unnecessary conflict.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-image/ios/ImageView.swift [hunk 1]
```diff
@@ -210,14 +210,28 @@ public final class ImageView: ExpoView {
...
       return
     }
+
+    // First we need to see if the image is animated. SDAnimatedImage might not properly render some image formats.
+    // If the image is animated, we want to use the data and create an SDAnimatedImage with that data.
+    // SDAnimatedImage extends UIImage so we should be able to do everything that we were doing before still.
+    var animated = false
+    var image = image
+
+    if image?.sd_isAnimated != nil, let data = data {
```
Comment (tsapeta) [inline packages/expo-image/ios/ImageView.swift:220]:
```suggestion
    if isAnimated, let data {
```

File: packages/expo-image/ios/ImageView.swift [hunk 2]
```diff
@@ -210,14 +210,28 @@ public final class ImageView: ExpoView {
...
+      animated = true
+      image = SDAnimatedImage(data: data)
+    }
```
Comment (tsapeta) [inline packages/expo-image/ios/ImageView.swift:223]: In general, maybe we can make a separate function for that, so we don't add too much logic in this `imageLoadCompleted` function? 🤔

File: packages/expo-image/ios/ImageView.swift [hunk 3]
```diff
@@ -210,14 +210,28 @@ public final class ImageView: ExpoView {
...
     if let image = image {
       onLoad([
         "cacheType": cacheTypeToString(cacheType),
         "source": [
           "url": imageUrl?.absoluteString,
           "width": image.size.width,
           "height": image.size.height,
-          "mediaType": imageFormatToMediaType(image.sd_imageFormat)
+          "mediaType": imageFormatToMediaType(image.sd_imageFormat),
+          "animated": animated
```
Comment (tsapeta) [inline packages/expo-image/ios/ImageView.swift:234]: For bool values, it feels more natural to use question-like names
```suggestion
          "isAnimated": isAnimated
```

File: packages/expo-image/src/Image.types.ts [hunk 4]
```diff
@@ -303,6 +311,8 @@ export interface ImageNativeProps extends ImageProps {
   placeholder?: ImageSource[];
   contentPosition?: ImageContentPositionObject;
   transition?: ImageTransition | null;
+  animated?: boolean;
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:314]: I guess it should be `autoplay`, right?
```suggestion
  autoplay?: boolean;
```

File: packages/expo-image/src/Image.types.ts [hunk 5]
```diff
@@ -437,6 +447,7 @@ export type ImageLoadEventData = {
     width: number;
     height: number;
     mediaType: string | null;
+    animated?: boolean | undefined;
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:450]: `undefined` is probably useless 🤔 

```suggestion
    isAnimated?: boolean;
```

**Review by tsapeta — CHANGES_REQUESTED** [review]
I'm impressed, it looks solid! 😍 Left just a few comments.
In terms of the implementation for Glide/Android, I hope @lukmccall could help with that.

File: apps/test-suite/tests/Image.js [hunk 6]
```diff
@@ -49,6 +53,18 @@ export async function test(t, { setPortalChild, cleanupPortal }) {
           }
         }
       });
+
+      if(Platform.OS === "ios") {
```
[the remark this answers is not in the document]
Comment (gkasdorf) [reply apps/test-suite/tests/Image.js:57]: Surprised the linter didn't pick that one up.

File: packages/expo-image/src/Image.types.ts (hunk 5 above)
Comment (gkasdorf) [reply packages/expo-image/src/Image.types.ts:450]: Yup!

File: packages/expo-image/src/Image.types.ts (hunk 4 above)
Comment (gkasdorf) [reply packages/expo-image/src/Image.types.ts:314]: Oops 😅 Looks like just the weird issue I love to throw into something!

**Comment by gkasdorf** [comment]
Thanks, cleaned up the minor issues and will fix up the ImageView a bit later today. Cheers.

File: apps/test-suite/tests/Image.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Simek) [reply apps/test-suite/tests/Image.js:57]: ESLint picked up this issue in CI, so it should do the same locally:
* https://github.com/expo/expo/actions/runs/6624239492/job/18000188472?pr=25008

LMK if something is not working for you correctly.

File: packages/expo-image/ios/ImageUtils.swift [hunk 7]
```diff
@@ -3,6 +3,19 @@
 import SDWebImage
 import ExpoModulesCore
 
+/**
+ Checks if the image is animated and returns an SDAnimatedImage if it does. Otherwise returns the UIImage.
+ */
+func createAnimatedIfNeeded(image: UIImage?, data: Data?) -> UIImage? {
```
Comment (gkasdorf) [inline packages/expo-image/ios/ImageUtils.swift:9]: Add a function to return an SDAnimatedImage if necessary inside of ImageUtils

File: docs/public/static/data/v49.0.0/expo-image.json
[the remark this answers is not in the document]
Comment (gkasdorf) [reply docs/public/static/data/v49.0.0/expo-image.json:1]: Done. Wasn't fully clear on the documentation updates, thanks for the heads up.

File: packages/expo-image/ios/ImageView.swift (hunk 2 above)
Comment (gkasdorf) [reply packages/expo-image/ios/ImageView.swift]: Added a function inside ImageUtils to do this.

File: packages/expo-image/ios/ImageView.swift [hunk 8]
```diff
@@ -210,14 +210,17 @@ public final class ImageView: ExpoView {
       log.debug("Loading the image has been canceled")
       return
     }
-    if let image = image {
+
+    // Create an SDAnimatedImage if needed then handle the image
+    if let image = createAnimatedIfNeeded(image: image, data: data) {
```
Comment (gkasdorf) [inline packages/expo-image/ios/ImageView.swift:215]: Here we either return an SDAnimatedImage or UIImage depending on whether the image is animated.

**Comment by gkasdorf** [comment]
@tsapeta Alright, that should clear up the issues. Let me know if you see anything else 👍

File: packages/expo-image/ios/ImageUtils.swift (hunk 7 above)
Comment (gkasdorf) [inline packages/expo-image/ios/ImageUtils.swift:9]: Add a function to return an SDAnimatedImage if necessary inside of ImageUtils

**Review by tsapeta — APPROVED** [review]
Looks perfect now, thank you! 🙏 Could you just resolve the conflicts? We merged another PR that caused a conflict 🙈

**Comment by gkasdorf** [comment]
@tsapeta Good to go now.!

**Comment by gkasdorf** [comment]
I’ll fix the test suite lint once other runs complete 👍

File: apps/test-suite/tests/Image.js [hunk 9]
```diff
@@ -57,6 +61,18 @@ export async function test(t, { setPortalChild, cleanupPortal }) {
...
+
+      if (Platform.OS === 'ios') {
+        t.it('load animated image and emits animated is true', async () => {
+          const event = await mountAndWaitFor(
+            <Image source={ANIMATED_IMAGE_SOURCE} style={{ height: 100, width: 100 }} />,
+            'onLoad',
+            setPortalChild
+          );
+
+          t.expect(event.source.animated).toBe(true);
```
Comment (tsapeta) [inline apps/test-suite/tests/Image.js:73]: Just one more thing 😂 
```suggestion
          t.expect(event.source.isAnimated).toBe(true);
```

**Review by tsapeta — APPROVED** [review]

**Comment by lukmccall** [comment]
> I am not overly familiar with Glide but am happy to implement the functionality on Android as well if anyone points me in the correct direction on that. If not, I will look more into it whenever I have some time.

Hi @gkasdorf, thanks for your contribution🏅 I think, I can give you some guidance on how to implement the same functionality on Android. In order to add the autoplay prop, you have to change only one `if` -  https://github.com/expo/expo/blob/714b55206e297ff857e4360119f263e67f53fd3e/packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageViewWrapper.kt#L363-L365
In this condition, we're starting the animation. 

To control the animation in the ExpoImageViewWrapper, you need to access the active view by invoking `ExpoImageViewWrapper.activeView`. From there, you can retrieve the current drawable by calling `.drawable`. If the drawable is an instance of `Animatable`, you can start or stop the animation.

**Comment by tsapeta** [comment]
This PR has been published in `1.7.0`. Thanks once again for your contribution! 🙏

**Comment by FelipeSSantos1** [comment]
@gkasdorf in some bigger animated images, I'm experiencing a slow-motion, did you experience that? it's easy to see that if we open the image on chrome and on iOS simulator or real device.
O Chrome it runs as a normal speed, unlike than iOS simulator/device.

Odd fact: if we view the image on Mac Finder preview the same issue happens!!!
it works: on chrome, using `react-native-fast-image` and converting the webp image to gif

https://github.com/expo/expo/assets/10479750/fd01a99f-0fee-4d38-90f7-6f17e418c47d

**Comment by tsapeta** [comment]
That's really odd behavior @FelipeSSantos1 🤔 Can you please open a new issue with this recording and provide a minimal reproducible example that includes this image?

**Comment by gkasdorf** [comment]
@FelipeSSantos1 When you put up a repro I'd be curious to look at it, mainly because I've actually had the *opposite* issue in the past: fast-image playing some animations two times faster instead of slower 😅 curious to test out that particular image 👍

**Comment by FelipeSSantos1** [comment]
https://github.com/expo/expo/issues/25920
thank you guys @tsapeta and @gkasdorf

**Pull request merged**